src/node/filesystem/binary.node.ts

```diff
@@ -0,0 +1,341 @@
+import { createReadStream } from 'fs';
+import type { FileHandle } from 'fs/promises';
+import { open } from 'fs/promises';
+import type { ReadableOptions } from 'stream';
+import { Readable } from 'stream';
+
+import type {
+  IBinaryData,
+  IBinaryFileMeta,
+  IDataContainer,
+  IDestructible,
+  IInitializable} from '../../interfaces';
+import {
+  isBinaryFileMeta,
+  isDestructible,
+  isInitializable,
+} from '../../interfaces';
+import { UnexpectedError } from '../../lib';
+import { handleNodeError } from './filesystem.node';
+
+export class StreamReader {
+  private stream: Readable;
+  private buffer: Buffer;
+  private ended = false;
+
+  private pendingReadResolve: (() => void) | undefined
+
+  constructor(stream: Readable) {
+    this.buffer = Buffer.from([]);
+    this.stream = stream;
+    this.stream.on('data', this.onData.bind(this));
+    this.stream.on('end', this.onEnd.bind(this));
+  }
+
+  private onData(chunk: Buffer) {
+    this.buffer = Buffer.concat([this.buffer, chunk]);
+    this.notifyData();
+  }
+
+  private onEnd() {
+    this.ended = true;
+    this.notifyData();
+  }
+
+  // assumption: this function is never called twice without awaiting its promise in between
+  private async waitForData(): Promise<void> {
+    this.stream.resume();
+
+    return new Promise((resolve, reject) => {
+      if (this.pendingReadResolve !== undefined) {
+        reject(new UnexpectedError('Waiting for data failed. Unable to resolve pending read'));
+      }
+      this.pendingReadResolve = resolve;
+    })
+  }
+
+  private notifyData() {
+    if (this.pendingReadResolve !== undefined) {
+      this.pendingReadResolve();
+    }
+
+    this.stream.pause();
+    this.pendingReadResolve = undefined;
+  }
+
+  public async read(size = 1): Promise<Buffer | undefined> {
+    while (this.buffer.length < size) {
+      if (this.ended) {
+        if (this.buffer.length > 0) {
+          // yield remaining data
+          break;
+        } else {
+          // signal EOF
+          return undefined;
+        }
+      }
+
+      await this.waitForData();
+    }
+
+    const chunk = this.buffer.subarray(0, size); // this might need to be a copy
+    this.buffer = this.buffer.subarray(size);
+
+    return chunk;
+  }
+}
+
+export class ReadableClone extends Readable {
+  constructor(private stream: Readable, options?: ReadableOptions) {
+    super(options);
+
+    this.stream.on('data', (chunk) => {
+      this.push(chunk);
```

Comment:
issue: I'm not sure about this, but I think this can be a big performance problems. The NodeJS docs contain a code sample which I think can be used here almost identically and it contains pausing mechanism, but that might interfere with point of this class.

https://nodejs.org/api/stream.html#readablepushchunk-encoding

Reply:
Ideally we would want something like this, but that's a web API - experimental and not the one we are using right now in Node https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/tee

Reply:
Oh, and that code is *not* trivial https://github.com/nodejs/node/blob/f9227ffc4b47da1ccc0b621001423f3221eca27a/lib/internal/webstreams/readablestream.js#L1451

Reply:
Thats what I wrote you about yesterday. And the code is comming from nodejs do documemtstion. Right now we need to either clone streamed data or getAllData to get data feom stream/file before sending it in request.

Anotger problem is, that if user consumes data from BinaryData and still passes BinaryData as payload it will be empty.